allow decoding byte string without padding

jad-elmourad opened the pull request:
### What was wrong?

closes #198

### How was it fixed?

@jad-elmourad:
- Allow decoding arbitrary length bytes without padding

@fselmo:
- Introduce this as a feature by allowing the default "strict mode" behavior for the ``BytesStringDecoder`` to be turned off via a ``strict`` flag that is ``True`` by default. The current behavior of the library is ``strict=True`` so this shouldn't be a breaking change. When ``strict=False``, the behavior desired in #198 is enabled.

- Add some testing around this and make sure all other decoding tests pass whether the ``strict`` flag is ``True`` or ``False``.

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://images.pexels.com/photos/50577/hedgehog-animal-baby-cute-50577.jpeg)

jad-elmourad committed:
do not require padding when decoding byte sting

jad-elmourad committed:
fix test: allow successful decoding with no padding

fselmo committed:
Add a strict flag to toggle decode 'strict mode'

fselmo reviewed (commented):
Hey @jad-elmourad. Thanks for getting this work started. This seems to be a desired feature for sure. I do think we should achieve this via a transient flag when calling `decode()` and when we retrieve all the decoders for each `TypeStr` in the `types`, we can set this transient flag on each of them before we return them. See my comment in #198.

Thoughts on that approach?

fselmo committed:
Some refactor + cleanup around strict flag; add tests

fselmo committed:
test both strict values for sanity check on all decode tests

fselmo committed:
Remove unnecessary checks for pad_size in tests

fselmo committed:
clean up byte-padding logic for strict vs non-strict

- Don't check anything beyond the data size for non-empty padding bytes with strict=False. This is the behavior in Solidity as well.
- f-strings instead of .format()

fselmo committed:
Add newsfragment & update docs for #198 / #211

reedsa reviewed (approved):
Looks good, ship it!

pacrob reviewed (approved):
lgtm!
